aip/general/0236.md

```diff
@@ -0,0 +1,355 @@
+  - If the Control Plane does not support a nested collection without a parent, the user **should** create a top-level resource which has the correct rules to represent a No-Op. Then, experiments can be added underneath that resource.
+- If multiple previews are supported, calling the create more than once on a resource **should** lead to the creation of multiple experiments represented as new subresources until the quota is reached.
+- If the operation failed, return the appropriate error code ([AIP-193][aip-193]).
+
+[aip-133]: https://aip.dev/133
+[aip-193]: https://aip.dev/193
+
+#### update
+
+```proto
+rpc UpdatePolicyExperiment(UpdatePolicyExperimentRequest) returns (PolicyExperiment);
+```
+- Standard methods: Update ([AIP-134]) **must** be used to change the fields-under-test in an experiment used for preview mode.
+  - The name inside “policy” **must not** be changed.
+- The Control Plane **should** set the PreviewMetadata.state to INACTIVE if the state was ACTIVE at the time of an update. This is so the user can easily distinguish between previews of different fields-under-test.
+- If the operation failed, the old experiment **must** be returned alongside the appropriate error code ([AIP-193][aip-193]).
+
+[aip-134]: https://aip.dev/134
+
+#### get
+
+```proto
+rpc GetPolicyExperiment(GetPolicyExperimentRequest) returns (PolicyExperiment);
+```
+- Standard methods: Get ([AIP-131][aip-131]) **must** be used to get the experiment used for preview mode.
+
+[aip-131]: https://aip.dev/131
+
+#### list
+
+```proto
+rpc ListPolicyExperiments(ListPolicyExperimentsRequest) returns (ListPolicyExperimentsResponse);
+```
+- Standard methods: List ([AIP-132][aip-132]) **should** be used to list experiments for preview mode.
+- List **should** support Filtering ([AIP-160][aip-160]) semantics.
+- Filtering on PreviewMetadata.state **should** indicate which experiments are actively being previewed or not.
+
+[aip-132]: https://aip.dev/132
+[aip-160]: https://aip.dev/160
+
+#### delete
+
+```proto
+rpc DeletePolicyExperiment(DeletePolicyExperimentRequest) returns (google.protobuf.Empty);
+```
+- Standard methods: Delete ([AIP-135][aip-135]) **must** be used to delete experiments.
+
+[aip-135]: https://aip.dev/135
+
+### Parent Resource API Changes
+
+#### delete
+
+- A delete of the top-level resource **must** delete all experiments.
+- To maintain the experiments, the top-level Policy **must** be changed to a no-op Policy instead of using delete.
+
+### Custom Method APIs
+
+#### startPreview
+
+```proto
+rpc StartPreviewPolicyExperiment(StartPreviewPolicyExperimentRequest) returns (StartPreviewPolicyExperimentResponse) {
+  option (google.api.http) = {
+    post: "/v1/{name=policies/*/policyExperiments/*}:startPreview"
+    body: "*"
+  }
+}
+
+message StartPreviewPolicyExperimentRequest {
+  string name = 1;
+  string etag = 2;
+}
+
+message StartPreviewPolicyExperimentResponse {
```

Comment:
Consider returning the PolicyExperiment (or LRO-of-PolicyExperiment) for custom methods whose results are effectively described by the state of the resource --- successful is implied by non-error result, and error_message is returned by Status error object.

The only new information in the response, parent_etag, may be a good one to add as an additional  PolicyExperiment string parent_etag [output_only] field.

That would be generally useful to know - what etag the parent had when the experiment was created.

Reply:
Done - however, the parent etag is actually already present inside the Policy message in the PolicyExperiment so we don't need to add it again separately. There's guidance earlier on line 88 saying the parent etag must be there.